# snapper

[BEAST 2](http://beast2.org) based package for SNP analyses -- difussion based [SNAPP](https://github.com/BEAST2-Dev/SNAPP).

## Installation

* Install BEAST 2 (available from [http://beast2.org](http://beast2.org)).
* Install snapper package through the [package manager](http://www.beast2.org/managing-packages/) (this automatically installs the SNAPP package as well)

## Tutorial
https://github.com/ForBioPhylogenomics/tutorials/tree/main/divergence_time_estimation_with_snp_data]

## Using snapper

* Start BEAUti, and select the snapper template  (meny `File/Templates/snapper`)
* Use BEAUti in a similar fashion as SNAPP (see (rough guide to SNAPP)[https://github.com/BEAST2-Dev/SNAPP/releases/download/v1.2.0/SNAPP.pdf] or the [SNAPP tutorial](https://github.com/BEAST2-Dev/beast-docs/releases/download/v1.0/SNAPP-tutorial-2018.zip))
* Run BEAST on the XML file saved from BEAUti

## Citing snapper

Marnus Stoltz, Boris Bauemer, Remco Bouckaert, Colin Fox, Gordon Hiscott, David Bryant.
Bayesian inference of species trees using diffusion models, Systematic Biology, 2020 DOI:10.1093/sysbio/syaa051

arXiv preprint [arXiv:1909.07276](https://arxiv.org/abs/1909.07276), 2019.


## Questions about snapper

BEAST user list: [https://groups.google.com/forum/#!forum/beast-users](https://groups.google.com/forum/#!forum/beast-users)

Marnus Stoltz: [stoltzstep@gmail.com](stoltzstep@gmail.com)

David Bryant: [david.bryant@otago.ac.nz](david.bryant@otago.ac.nz)
